### PR TITLE
Cleanup of CMake and Docker files

### PR DESCRIPTION
**Pull request by stv0g**
This PR resolves several inconsistencies in the CMakeFiles.txt and Dockerfiles of the project which have been accumulating over the last years. It is required to build DPsim on OPAL-RT Targets. With the PR we also update the Version of VILLAS and the Fedora version to 36.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
-cmake_minimum_required(VERSION 3.12...3.23)
+# Please keep 3.14 as minimum requirement for building on OPAL-RT targets
+cmake_minimum_required(VERSION 3.14...3.23)
```
Comment (n-eiling) [inline CMakeLists.txt:2]: I don't understand why this should be necessary. 3.14 is between 3.12 and 3.23.
This is a change I would really rather avoid, because it for sure breaks compiling on certain targets.

File: CMakeLists.txt [hunk 2]
```diff
@@ -66,161 +68,175 @@ check_symbol_exists(getopt_long getopt.h HAVE_GETOPT)
...
 add_subdirectory(docs)
 
-include(cmake/GetGridData.cmake)
+if(WITH_VILLAS)
+	add_subdirectory(dpsim-villas)
+endif()
 
-message(VERBOSE "Compiler Release Flags: " 	${CMAKE_CXX_FLAGS_RELEASE})
-message(VERBOSE "Compiler Debug Flags: " 	${CMAKE_CXX_FLAGS_DEBUG})
+if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.16")
```
Comment (n-eiling) [inline CMakeLists.txt:218]: why is this necessary? Was this wrong before?

File: cmake/GetVersion.cmake [hunk 3]
```diff
@@ -1,95 +1,95 @@
...
-# ${PREFIX}_GIT_REV           1cd25c2f
-# ${PREFIX}_GIT_BRANCH        ci-tests
+# ${PREFIX}_VERSION_STR	   v0.6.3
+# ${PREFIX}_VERSION		   0.6.3
+# ${PREFIX}_MAJOR_VERSION	 0
+# ${PREFIX}_MINOR_VERSION	 6
+# ${PREFIX}_PATCH_VERSION	 3
+# ${PREFIX}_RELEASE		   1.ci_tests_release.20180823git1cd25c2f
+# ${PREFIX}_GIT_REV		   1cd25c2f
+# ${PREFIX}_GIT_BRANCH		ci-tests
```
Comment (n-eiling) [inline cmake/GetVersion.cmake:10]: I think the old version is prettier. This change seems bad.

**Review by n-eiling — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 4]
```diff
@@ -66,161 +68,175 @@ check_symbol_exists(getopt_long getopt.h HAVE_GETOPT)
...
-option(WITH_ASAN "Adds compiler flags to use the address sanitizer" OFF)
-option(WITH_TSAN "Adds compiler flags to use the thread sanitizer" OFF)
-option(CGMES_BUILD "Build with CGMES instead of CIM" OFF)
+# Build options
+option(FETCH_EIGEN          "Fetch Eigen3 as module" OFF)
+option(FETCH_SPDLOG         "Fetch spdlog as module" OFF)
+option(FETCH_CIMPP          "Fetch CIMpp as module" OFF)
+option(FETCH_PYBIND         "Fetch pybind11 as module" OFF)
+option(FETCH_GRID_DATA      "Fetch grid data" ON)
+option(FETCH_FILESYSTEM     "Fetch standalone implememtation of std::filesystem" OFF)
```
[the remark this answers is not in the document]
Comment (stv0g) [reply CMakeLists.txt:77]: Yes, thanks for pointing it out. Its definitely something to discuss.

Here is my motivation for updating it:
- The old naming is wrong. `FetchContent` modules are not submodules
- Its inconsistent: `MODULE` vs `SUBMODULE`.
- `Fetch` is more aligned with what it actually does: Fetching external dependncies.

File: CMakeLists.txt (hunk 2 above)
Comment (stv0g) [reply CMakeLists.txt:218]: The `VERBOSE` message level has been introduced by CMake 3.16 and hence is not supported with earlier CMake versions.

File: cmake/FetchCIMpp.cmake [hunk 5]
```diff
@@ -8,26 +6,17 @@ message(STATUS "CIM Version: ${CIM_VERSION}")
...
+	set(CGMES_BUILD ON)
 endif()
 
 include(FetchContent)
-FetchContent_Declare(
-  cimpp-module
-  GIT_REPOSITORY https://github.com/cim-iec/libcimpp.git
-  GIT_TAG        master
+FetchContent_Declare(cimpp-module
+	GIT_REPOSITORY https://github.com/cim-iec/libcimpp.git
```
[the remark this answers is not in the document]
Comment (stv0g) [reply cmake/FetchCIMpp.cmake:16]: Because its redundant. FetchContent will fetch the default branch which is currently `master`.
The old approach will fail if somebody renames the default branch.

File: cmake/GetVersion.cmake (hunk 3 above)
Comment (stv0g) [reply cmake/GetVersion.cmake:10]: Agreed. I will fix it. My editor messed it up again :(

File: CMakeLists.txt (hunk 1 above)
Comment (stv0g) [reply CMakeLists.txt:2]: 3.14 is the first version which supports `FetchContent_MakeAvailable` and hence simplifies the `cmake/Fetch*.cmake` CMake files a lot.

At the same time, its the lowest version of CMake which is shipped by the supported Ubuntu, Debian, Rocky and Fedora Linux distros.

Actually 3.14 is the version shipped with OPAL RT-Linux. The other Linux distros ship at least with 3.16

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (stv0g) [reply CMakeLists.txt:77]: I also checked and updated for any use of the old naming in the documentation and Bash scripts / CI config.

File: CMakeLists.txt (hunk 1 above)
Comment (n-eiling) [reply CMakeLists.txt:2]: ok sounds fine then

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (n-eiling) [reply CMakeLists.txt:77]: Can we also include the old ones and add a deprecated warning or something similar? If yes, then I think we can merge this without much issues. If this is not possible, we should really test this in-depth and also have instructions on how to update an existing repo.

[the remark this answers is not in the document]
Comment (stv0g) [reply CMakeLists.txt:86]: I have kept the old options now and added the new ones which take their default value from the old options..

I will also create a new issue to cleanup the deprecated options some time in the future.

**Comment by stv0g** [comment]
@dinkelbachjan @n-eiling

I pushed more changes to this PR. Lets see if the CI is also happy.

File: dpsim/examples/cxx/cim_graphviz/CMakeLists.txt [hunk 6]
```diff
@@ -1,4 +1,4 @@
 if(WITH_GRAPHVIZ AND WITH_CIM)
```
Comment (dinkelbachjan) [inline dpsim/examples/cxx/cim_graphviz/CMakeLists.txt:1]: Shouldn't `WITH_CIM` then be removed here, if `CIMPP_LIBRARIES` not anymore considered below?

File: packaging/Docker/Dockerfile.dev [hunk 7]
```diff
@@ -1,11 +1,16 @@
-FROM fedora:34 AS base
+FROM fedora:36 AS base
+
+ARG CIM_VERSION=CGMES_2.4.15_16FEB2016
+ARG VILLAS_VERSION=c976cd62d8c6667a078be0785ca3e623a05f7456
```
Comment (dinkelbachjan) [inline packaging/Docker/Dockerfile.dev:4]: You updated VILLASnode to a later commit. Any specific reason for that? Can you mention this also in PR description (along with upgrade of Dockerfile to Fedora 36)?

File: cmake/GetVersion.cmake [hunk 8]
```diff
@@ -1,95 +1,95 @@
...
 # ${PREFIX}_GIT_BRANCH_NORM   ci_tests
-# ${PREFIX}_VARIANT           release
-# ${PREFIX}_VARIANT_NORM      release
-# ${PREFIX}_BUILD_ID          v0.6.3-1cd25c2f-release
-# ${PREFIX}_BUILD_DATE        20180823
+# ${PREFIX}_VARIANT		   release
+# ${PREFIX}_VARIANT_NORM	  release
+# ${PREFIX}_BUILD_ID		  v0.6.3-1cd25c2f-release
+# ${PREFIX}_BUILD_DATE		20180823
```
Comment (n-eiling) [inline cmake/GetVersion.cmake:15]: formatting

File: cmake/GetVersion.cmake [hunk 9]
```diff
@@ -1,95 +1,95 @@
...
+		set(${PREFIX}_MINOR_VERSION   ${MINOR_VERSION}	PARENT_SCOPE)
+		set(${PREFIX}_PATCH_VERSION   ${PATCH_VERSION}	PARENT_SCOPE)
+		set(${PREFIX}_RELEASE		 ${RELEASE}		  PARENT_SCOPE)
+		set(${PREFIX}_GIT_REV		 ${GIT_REV}		  PARENT_SCOPE)
+		set(${PREFIX}_GIT_BRANCH	  ${GIT_BRANCH}	   PARENT_SCOPE)
+		set(${PREFIX}_GIT_BRANCH_NORM ${GIT_BRANCH_NORM}  PARENT_SCOPE)
+		set(${PREFIX}_VARIANT		 ${VARIANT}		  PARENT_SCOPE)
+		set(${PREFIX}_VARIANT_NORM	${VARIANT_NORM}	 PARENT_SCOPE)
+		set(${PREFIX}_BUILD_ID		${BUILD_ID}		 PARENT_SCOPE)
+		set(${PREFIX}_BUILD_DATE	  ${BUILD_DATE}	   PARENT_SCOPE)
```
Comment (n-eiling) [inline cmake/GetVersion.cmake:93]: formatting was better before

File: dpsim-models/CMakeLists.txt [hunk 10]
```diff
@@ -4,31 +4,15 @@ configure_file(
 )
 
 set(MODELS_INCLUDE_DIRS
-	${EIGEN3_INCLUDE_DIR}
-	${SPDLOG_INCLUDE_DIR}
+	${CMAKE_CURRENT_SOURCE_DIR}/include
+	${CMAKE_CURRENT_BINARY_DIR}/include
```
Comment (n-eiling) [inline dpsim-models/CMakeLists.txt:8]: this change seems weird

**Review by n-eiling — COMMENTED** [review]
Do you think it would be useful to split change PR and formatting PR and not do both in a single one? This would make reviewing way easier and probably also rebasing.

File: cmake/FetchFilesystem.cmake [hunk 11]
```diff
@@ -0,0 +1,13 @@
+include(FetchContent)
+FetchContent_Declare(filesystem-module
+	GIT_REPOSITORY https://github.com/gulrak/filesystem
```
[the remark this answers is not in the document]
Comment (stv0g) [reply cmake/FetchFilesystem.cmake:3]: Its an optional dependency only required for systems which come without libstdc++fs.
I think its a rare case that this is needed (e.g. OPAL-RT targets).

Should I still document it?

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (stv0g) [reply CMakeLists.txt:87]: Here is the issue: https://github.com/sogno-platform/dpsim/issues/187

File: packaging/Docker/Dockerfile.dev (hunk 7 above)
Comment (stv0g) [reply packaging/Docker/Dockerfile.dev:4]: The new VILLASnode version contains fixes which I needed for building it on OPAL-RT targets.

I added a comment to the PR description.

File: CMakeLists.txt [hunk 12]
```diff
@@ -66,161 +70,187 @@ check_symbol_exists(getopt_long getopt.h HAVE_GETOPT)
...
-option(WITH_ASAN "Adds compiler flags to use the address sanitizer" OFF)
-option(WITH_TSAN "Adds compiler flags to use the thread sanitizer" OFF)
-option(CGMES_BUILD "Build with CGMES instead of CIM" OFF)
+# Deprecated build options
+# TODO: Remove deprecated options in future release
+option(WITH_EIGEN_MODULE     "Fetch Eigen3 as module (deprecated option)" OFF)
+option(WITH_SPDLOG_SUBMODULE "Fetch spdlog as module (deprecated option)" OFF)
+option(WITH_CIM_SUBMODULE    "Fetch CIMpp as module (deprecated option)" OFF)
+option(WITH_PYBIND_MODULE    "Fetch pybind11 as module (deprecated option)" OFF)
+option(GET_GRID_DATA     	   "Fetch grid data (deprecated option)" ON)
```
[the remark this answers is not in the document]
Comment (stv0g) [reply CMakeLists.txt:79]: Fixed

File: cmake/GetVersion.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (stv0g) [reply cmake/GetVersion.cmake:10]: Fixed

File: cmake/GetVersion.cmake (hunk 8 above)
Comment (stv0g) [reply cmake/GetVersion.cmake:15]: Fixed

File: cmake/GetVersion.cmake (hunk 9 above)
Comment (stv0g) [reply cmake/GetVersion.cmake:93]: Fixed

File: dpsim-models/CMakeLists.txt (hunk 10 above)
Comment (stv0g) [reply dpsim-models/CMakeLists.txt:8]: This are actually two unrelated changes:

- The spdlog and Eigen include directories are not required here as they are gathered by the libraries (line 14-15)
- We add the included directories local to this target

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -1,11 +1,14 @@
...
   GIT_STRATEGY: fetch
   GIT_SUBMODULE_STRATEGY: recursive
+
   RSYNC_OPTS: --recursive --ignore-missing-args --chown ${DEPLOY_USER}:${DEPLOY_USER}
+
   DOCKER_IMAGE: sogno/dpsim
   DOCKER_IMAGE_DEV: sogno/dpsim-dev
   DOCKER_IMAGE_DEV_ROCKY: sogno/dpsim-dev-rocky
 
+  MAKE_OPTS: -j16 # for HYPER buildbots
```
[the remark this answers is not in the document]
Comment (stv0g) [reply .gitlab-ci.yml:11]: The old builds where already made in parallel:

https://github.com/sogno-platform/dpsim/pull/168/files#diff-037ea159eb0a7cb0ac23b851e66bee30fb838ee8d0d99fa331a1ba65283d37f7L117

I now just hard code the parallel level here so we dont overwhelm the CI runner and still leave some resources to other users.

File: dpsim/examples/cxx/cim_graphviz/CMakeLists.txt (hunk 6 above)
Comment (stv0g) [reply dpsim/examples/cxx/cim_graphviz/CMakeLists.txt:1]: I think we should probably also link against cimpp here. I added it as a library dependency.

File: cmake/FetchCIMpp.cmake [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+elseif(CIM_VERSION STREQUAL "CGMES_2.4.15_16FEB2016")
+	set(USE_CIM_VERSION "CGMES_2.4.15_16FEB2016")
+	set(CGMES_BUILD ON)
+endif()
+
+set(CIMPP_BUILD_DOC OFF)
+
+include(FetchContent)
+FetchContent_Declare(cimpp-module
+	GIT_REPOSITORY https://github.com/stv0g/libcimpp.git
```
[the remark this answers is not in the document]
Comment (stv0g) [reply cmake/FetchCIMpp.cmake]: Oh sorry. This was only intended to be in place as long the PR (https://github.com/sogno-platform/libcimpp/pull/21) in libcimpp has not been merged.

I reverted it back to the sogno-platform repo.

File: .gitlab-ci.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (n-eiling) [reply .gitlab-ci.yml:11]: hm okay

File: cmake/FetchFilesystem.cmake (hunk 11 above)
[the remark this answers is not in the document]
Comment (n-eiling) [reply cmake/FetchFilesystem.cmake:3]: I mean this is a really small dependency but it was not clear that you added one. I think we should be at least a bit caution to not add weird licenses or huge inactive code bases. Seems to be no issue here though.

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,33 +1,37 @@
...
-  cmake_policy(SET CMP0076 NEW)
+	cmake_policy(SET CMP0076 NEW)
 endif()
 # Allow setting Graphviz_ROOT for windows builds
 if(WIN32)
-    cmake_policy(SET CMP0074 NEW)
+	cmake_policy(SET CMP0074 NEW)
 endif()
 
+set(CMAKE_POLICY_DEFAULT_CMP0077 NEW)
```
[the remark this answers is not in the document]
Comment (n-eiling) [reply CMakeLists.txt:28]: and for what exactly do you need it?

**Review by dinkelbachjan — APPROVED** [review]

**Review by n-eiling — APPROVED** [review]
